Port redmine2 cookbook to transportapi's infrastructure

madwort opened the pull request:
- Use Rackspace for kitchen tests
- Use transportapi-rbenv cookbook instead of aminin/chef-rbenv
- use redmine-3.2.0 with ruby 2.3.0
- Use bundle to run rake when installing plugins
- Allow root SSH access for kitchen
- Use latest chef & smallest rackspace machine
- Update gem dependencies
- Fix SSL certificate kitchen tests & spec tests from upstream
- Configure CircleCI to use rackspace for kitchen tests

Commit:
Port redmine2 cookbook to transportapi's infrastructure

* Use Rackspace for kitchen tests
* Use transportapi-rbenv cookbook instead of aminin/chef-rbenv
* use redmine-3.2.0 with ruby 2.3.0
* Use bundle to run rake when installing plugins

madwort commented:
`aminin/chef-rbenv` is an alternative cookbook for providing `rbenv` to the one that transportapi is using, with a different API, but it shares the same cookbook name (just `rbenv`). This means that it was impossible to deploy using this redmine2 cookbook without porting it to our `rbenv` cookbook as this PR does.

mgogov commented on .kitchen.yml:
```diff
@@ -1,22 +1,19 @@
+  rackspace_username: <%= ENV['RACKSPACE_USERNAME'] %>
+  rackspace_api_key: <%= ENV['RACKSPACE_API_KEY'] %>
+  rackspace_region: 'lon'
+  flavor_id: 'performance1-8'
```
Isn't `performance1-1` sufficient enough?

mgogov commented on .kitchen.yml:
```diff
@@ -1,22 +1,19 @@
+  rackspace_region: 'lon'
+  flavor_id: 'performance1-8'
+  privileged: true
+  require_chef_omnibus: 12.5.1
```
Hm, most cookbooks now use `12.7.2`, maybe we should use that?

mgogov commented on .kitchen.yml:
```diff
@@ -55,12 +52,18 @@ suites:
         server_debian_password: q654321
   - name: postgresql
     run_list:
+      - recipe[apt]
```
Um.. shouldn't we use our `transportapi-apt` wrapper cookbook over the original `apt-cookbook`? Not sure what's the intent here but thought I'd raise it so we're sure to pick the right one.

madwort replied:
good question. Just tested it & actually it's not required here, so removing.

madwort replied:
removed as they're not needed to run the kitchen tests on rackspace.

mgogov commented on attributes/default.rb:
```diff
@@ -3,8 +3,8 @@
-default[:redmine][:version]      = '2.6.1'
-default[:redmine][:ruby_version] = '1.9.3-p484'
+default[:redmine][:version]      = '3.2.0'
+default[:redmine][:ruby_version] = '2.3.0'
```
Just to be on the safe side, I suppose that redmine version plays well with that ruby version?

madwort replied:
Yep, main issue was finding a combination that worked with the `redmine_helpdesk` plugin - see thread here: https://github.com/jfqd/redmine_helpdesk/issues/110#issuecomment-181966153

mgogov commented on recipes/default.rb:
```diff
@@ -34,28 +34,26 @@
-  rake_command = 'rake'
-  ruby_command = 'ruby'
+  rake_command = 'bundle exec rake'
+  ruby_command = 'bundle exec ruby'
```
👍

mgogov commented on recipes/default.rb:
```diff
@@ -34,28 +34,26 @@
+
+  bundle_command = "/opt/rbenv/shims/bundle"
+  rake_command = "#{bundle_command} exec /opt/rbenv/shims/rake"
+  ruby_command = "#{bundle_command} exec /opt/rbenv/shims/ruby"
```
Hm, just wondering whether it would make sense to reuse the `ruby-cookbook` here since it sets up most of this stuff already.
Pros:
- already exists and works and we'll reuse stuff
- does some additional thinks like pinning the bundle / rbenv version, etc
  Cons:
- might deviate from the setup here
- might not do all of the setup being done here

So, I guess it's a no-go for now but thought I'd raise it to briefly discuss it.

Commit:
Allow root SSH access for kitchen

Commit:
Use latest chef & smallest rackspace machine

Commit:
Update gem dependencies

Commit:
Remove unneccesary items from postgres kitchen runlist

Commit:
Fix SSL certificate kitchen tests

Commit:
Fix spec tests

madwort commented:
Ok, think I've addressed all your comments, also I've fixed the tests so the serverspec tests work (I think they were broken upstream for `postgres` - ported some bits from the `mysql` tests).

Commit:
Initial CircleCI config

Commit:
Add .ruby-version (fix CircleCI)

Commit:
Add `kitchen-docker` for CircleCI compatibility

Commit:
Update Gemfile.lock (might fix CircleCI issue)

Commit:
Add an insanely high timeout (200mins) to see if this fixes our build

mgogov commented:
`command bundle exec kitchen test -c 4 --destroy=always took more than 120 minutes to run` and build failed again ):

madwort commented:
thanks, will investigate

madwort commented:
ok, running it in a `kitchen-rackspace` instance, next thing that happens is 

```
             * ruby_block[wait for redmine service socket] action run
        - execute the ruby block wait for redmine service socket
             - execute the ruby block restart_service
           * ruby_block[restart_log_service] action run[2016-05-25T09:07:59+00:00]
```

so maybe this won't work in a docker instance as it's a service? cc: @blaskov 
Doing a CI run with SSH to investigate...

madwort commented:
Nah, that's hogwash, it restarts the service earlier in the converge

Commit:
Use rackspace instead of docker for CI tests

* bad thing: more server costs
* good thing: it might actually work

madwort commented:
So, haven't been able to get it working in a docker container, but it works on rackspace. Can we merge it like this? We hopefully won't update this cookbook very often...

madwort commented:
FAO @mgogov

blaskov commented on .kitchen.ci.yml:
```diff
@@ -0,0 +1,48 @@
+
+suites:
+  - name: postgresql
+    data_bags_path: "test/integration/mysql/data_bags"
```
The test suite is named `postgresql`, but we have MySQL-related data bags? :-)

madwort replied:
Well spotted. So, in the upstream cookbook, the `mysql` suite had MySQL-related data bags, and the `postgres` suite was broken because it had no data bags, so I re-used the supplied MySQL ones. They're actually generic data so should be in a generic place. I'll do that now...

meddle0x53 replied:
Hmm why `mysql` data bag if the suite is `postgres`?

blaskov commented on .kitchen.ci.yml:
```diff
@@ -0,0 +1,48 @@
+    data_bags_path: "test/integration/mysql/data_bags"
+    encrypted_data_bag_secret_key_path: "test/integration/mysql/data_bag_key"
+    run_list:
+      - recipe[iptables]
```
👍 
(We should probably use this on more servers!)

blaskov replied:
Oh, it's only for the CI test suite and not used anywhere else. Anyway, above statement still stands :-)

meddle0x53 commented:
Happened to me with different cookbooks/recipes a few times too. Some things just can't be run with CircleCI/docker... For example `upstart` is not running with CircleCI/docker so a lot of services are not running too.

meddle0x53 commented on .kitchen.yml:
```diff
@@ -54,19 +51,30 @@ suites:
         server_repl_password: q654321
         server_debian_password: q654321
   - name: postgresql
+    data_bags_path: "test/integration/mysql/data_bags"
```
As above^

Commit:
Use correct set of databags

blaskov commented on .kitchen.yml:
```diff
@@ -54,19 +51,30 @@ suites:
         server_repl_password: q654321
```
Can't comment on above lines for some reason, but do we really need these MySQL settings?

madwort replied:
Hmm, my plan with this was to patch the postgresql support to suit our needs and not touch the mysql support. I could remove these tests if you'd rather? To be clear, this is a compatibility patch to an upstream application cookbook, we'll have a wrapper cookbook for this as well with our own config in.

madwort replied:
(you can't comment on the lines above because they're from the current version of the upstream cookbook, not from this PR)

blaskov replied:
Let's keep it as is then 👍

blaskov commented:
This is a merge from me, once the tests pass and I get answer to my last question (even if it's "Let's leave it as is").
